File: py/portfolio.py
import functools
import operator
# 
from money import Money

class Portfolio:
    def __init__(self):
        self.moneys = []
    
    def add(self, *moneys):
        self.moneys.extend(moneys)
    
    def evaluate(self, bank, currency):
        total = 0
        failures = []

        for m in self.moneys:
            try:
                total += bank.convert(m, currency).amount
            except Exception as ex: # there is no matched conversion
                failures.append(str(ex.args[0]))

        if len(failures) > 0: # there is an error
            failureMessage = ','.join(failures)
            raise Exception(f"Missing exchange rate(s):[{failureMessage}]")

        return Money(total, currency)
    
    def __convert(self, money, currency):
        if money.currency == currency:
            return money.amount

        exchangeRates = {
            'EUR->USD': 1.2,
            'USD->KRW': 1100,
        }

        key = f'{money.currency}->{currency}'
        return money.amount * exchangeRates[key]

File: go/stocks/portfolio.go
package stocks

import "errors"

// Author:	DragonChen https://github.com/dragonchen-tw/
// Title:	portfolio package
// Date:	2022/08/30

type Portfolio []Money

func (p Portfolio) Add(money Money) Portfolio {
	p = append(p, money)
	return p
}

func (p Portfolio) Evaluate(bank Bank, currency string) (*Money, error) {
	total := 0.0
	failedConversions := make([]string, 0)
	for _, m := range p {
		if converterMoney, ok := bank.Convert(m, currency); ok != nil {
			failedConversions = append(failedConversions, m.currency+"->"+currency)
		} else {
			total += converterMoney.amount
		}
	}

	if len(failedConversions) != 0 {
		failures := "["
		for _, f := range failedConversions {
			failures += f + ","
		}
		failures += "]"

		return nil, errors.New("Missing exchange rate(s):" + failures)
	}

	var result Money
	result = NewMoney(total, currency)
	return &result, nil
}

func convert(m Money, currency string) (float64, bool) {
	exchangeRates := map[string]float64{
		"EUR->USD": 1.2,
		"USD->KRW": 1100,
	}

	if m.currency == currency {
		return m.amount, true
	}
	key := m.currency + "->" + currency
	rate, ok := exchangeRates[key]
	return m.amount * rate, ok
}
